src/login/Login.cpp

```diff
@@ -267,3 +354,1805 @@ void ProfileRecord::FormatTo(char* buffer, size_t length) const
 
 	*result.out = 0;
 }
+
+class WithDb
+{
+public:
+	explicit WithDb(const int flags)
+	{
+		if (sqlite3_open_v2(s_dbPath.c_str(), &m_db, flags, nullptr) != SQLITE_OK)
+		{
+			SPDLOG_ERROR("AutoLogin Error failed to open database %s: %s", s_dbPath, sqlite3_errmsg(m_db));
+			sqlite3_close(m_db);
+			m_db = nullptr;
+		}
+		else
+		{
+			// pragmas don't need return values, just execute them
+			sqlite3_exec(m_db, "PRAGMA journal_mode = WAL", nullptr, nullptr, nullptr);
+			sqlite3_exec(m_db, "PRAGMA synchronous = normal", nullptr, nullptr, nullptr);
+			sqlite3_exec(m_db, "PRAGMA temp_store = memory", nullptr, nullptr, nullptr);
+			sqlite3_exec(m_db, "PRAGMA mmap_size = 100000000", nullptr, nullptr, nullptr);
+			sqlite3_exec(m_db, "PRAGMA foreign_keys = ON", nullptr, nullptr, nullptr);
+		}
+	}
+
+	~WithDb()
+	{
+		if (m_db != nullptr) sqlite3_close(m_db);
+	}
+
+	[[nodiscard]] sqlite3* GetDB() const { return m_db; }
+
+	template <typename T>
+	static std::function<T(const std::string&, const std::function<T(sqlite3_stmt*, sqlite3*)>&)> Query(int flags)
+	{
+		return [flags](const std::string& query, const std::function<T(sqlite3_stmt*, sqlite3*)>& action)
+			{
+				const auto connection = Get(flags);
+				return WithStatement<T>(connection->m_db, query)(action);
+			};
+	}
+
+	WithDb(const WithDb&) = delete;
+	WithDb(WithDb&&) = delete;
+	WithDb& operator=(const WithDb&) = delete;
+	WithDb& operator=(WithDb&&) = delete;
+
+	static const std::shared_ptr<WithDb>& Get(int flags)
+	{
+		static std::map<int, std::shared_ptr<WithDb>> connections;
```

Comment:
how do you remove connections?

Reply:
they get shut down gracefully when the object dies, but I'll add a Close static to do that

Reply:
its a shared_ptr stored in a function-level static. it will never die.